### ISSUE TITLE
Update to Pydantic 2 and FastAPI 0.100.0

### ISSUE THREAD

**Issue by mattt** [replicate/cog#1186]
From https://fastapi.tiangolo.com/release-notes/#01000:

> Pydantic version 2 has the core re-written in Rust and includes a lot of improvements and features... In some cases, for pure data validation and processing, you can get performance improvements of 20x or more.

**Comment by nickstenning** [comment]
Cross-linking this comment from the dependabot PR so we don't lose it: https://github.com/replicate/cog/pull/1187#issuecomment-1631174754

**Comment by mattt** [comment]
Related to #1216

**Comment by mattkindy** [comment]
This will prevent use of packages (e.g. https://github.com/outlines-dev/outlines ) that have dependencies on pydantic and have upgraded to v2 in their most recent updates.

**Comment by sinopec** [comment]
really need it ,specially when some  llm inference engine requires pydantic 2....
annoying conflicts.....

**Comment by dkhokhlov** [comment]
> really need it ,specially when some llm inference engine requires pydantic 2.... annoying conflicts.....
@sinopec - could you give model examples for testing?

**Comment by jd7h** [comment]
I deployed a model today that required pydantic==2.5.3. While we're waiting for the update to newer versions of pydantic and fastAPI, here are some possible workarounds. Choose one depending on the speed you require and how tightly integrated your code is with the incompatible versions of pydantic and fastAPI:

## Supplemental venv
Create a virtual environment with venv to supplement the packages installed from `cog.yaml`
```
$ cog run bash
# python -m venv --system-site-packages venv_pydantic
# source venv_pydantic/bin/activate
# pip install pydantic==2.5.3
```
predict.py:
```
import subprocess
subprocess.check_call(["venv_pydantic/bin/python", "this_script_requires_pydantic2"])
```
Drawback: You can't easily pre-load your modelweights in `setup()`.

## Poetry
2. in cog.yaml, git clone your code and install it with poetry
cog.yaml:
```
python_packages:
  - poetry

run:
  - git clone myrepo.git && cd myrepo && poetry install
```
in predict.py:
```
import subprocess
subprocess.check_call("cd /myrepo && python -m poetry run python -m script.py".split())
```
Drawback: You can't easily pre-load your modelweights in `setup()`.

## Monkeypatching
3. You can Mock() or otherwise monkeypatch parts of incompatible libraries. 
predict.py:
```
from unittest.mock import Mock

sys.modules['pydantic.computed_field'] = Mock()
sys.modules['pydantic_settings'] = Mock()
sys.modules['pydantic_settings'].BaseSettings = MyCustomSettings
```
This hack works best if the pydantic/fastAPI modules are imported but not essential for your code. The advantage is that you can use setup() to preload model weights.
For a more elaborate example see the [texify cog demo](https://github.com/datakami-models/texify/blob/79221f0708671ae85fa8145a7faf9c547e03125a/predict.py#L19-L42). I used this method for texify because this way I didn't need to modify the original model code to make the cog demo work.

**Comment by yorickvP** [comment]
Summarizing here:
- Upgrading pydantic also requires a fastAPI upgrade, which will bump the generated OpenAPI schemas from 3.0 to 3.1.
- kin-openapi, which we use in cog and registry-proxy (potentially other places), does not support this: https://github.com/getkin/kin-openapi/issues/230, and has asked for sponsorship to update it
- I attempted to switch to libopenapi: https://github.com/replicate/cog/pull/1539, but concluded its validation is not as good as kin-openapi

cc @nickstenning for the sponsorship
cc @dkhokhlov for coordination
Another issue for this: https://github.com/replicate/cog/issues/1336

**Comment by nickstenning** [comment]
I chatted to @yorickvP on Slack. Here's a quick summary of what we talked about:

- I'm very happy in principle for us to sponsor `kin-openapi` but we'd need to contact the maintainer and understand whether that would actually get us what we need. The linked issue about sponsorship is several years old.
- In practice we use `kin-openapi` with Replicate's own 3.1 spec without issue. It seems likely we can get away with it for most if not all Cog models.
- We have a decent little library of test models/fixtures in this repo that we can use to test that hypothesis.
